src/mono/mono/utils/memfuncs.c

```diff
@@ -292,4 +292,75 @@ mono_determine_physical_ram_size (void)
 #else
 	return 134217728;
 #endif
-}
\ No newline at end of file
+}
+
+guint64
+mono_determine_physical_ram_available_size (void)
+{
+#if defined (TARGET_WIN32)
+	MEMORYSTATUSEX memstat;
+
+	memstat.dwLength = sizeof (memstat);
+	GlobalMemoryStatusEx (&memstat);
+	return (guint64)memstat.ullAvailPhys;
+
+#elif defined (__NetBSD__)
+	struct vmtotal vm_total;
+	guint64 page_size;
+	int mib[2];
+	size_t len;
+
+	mib[0] = CTL_VM;
+	mib[1] = VM_METER;
+
+	len = sizeof (vm_total);
+	sysctl (mib, 2, &vm_total, &len, NULL, 0);
+
+	mib[0] = CTL_HW;
+	mib[1] = HW_PAGESIZE;
+
+	len = sizeof (page_size);
+	sysctl (mib, 2, &page_size, &len, NULL, 0);
+
+	return ((guint64) vm_total.t_free * page_size) / 1024;
+#elif defined (__APPLE__)
+	mach_msg_type_number_t count = HOST_VM_INFO_COUNT;
+	mach_port_t host = mach_host_self ();
+	vm_size_t page_size;
+	vm_statistics_data_t vmstat;
+	kern_return_t ret;
+	do {
+		ret = host_statistics (host, HOST_VM_INFO, (host_info_t)&vmstat, &count);
```

Comment:
I guess we're okay using the 32-bit version of this here since the returned value is in pages rather than bytes?

Reply:
I believe that is correct.